xdem/spatial_tools.py

```diff
@@ -398,6 +408,149 @@ def hillshade(dem: Union[np.ndarray, np.ma.masked_array], resolution: Union[floa
+    :param margin_improvement: improvement margin (percentage) below which the lesser degree polynomial is kept
+    :param linear_pkg: package to use for Linear estimator, one of 'scipy' and 'sklearn'
+    :param random_state: random seed for testing purposes
+    :param verbose: if text should be printed
+
+    :returns coefs, degree: polynomial coefficients and degree for the best-fit polynomial
+    """
+    if not isinstance(estimator, str) or estimator not in ['Linear','Theil-Sen','RANSAC','Huber']:
+        raise ValueError('Attribute estimator must be one of "Linear", "Theil-Sen", "RANSAC" or "Huber".')
+    if not isinstance(linear_pkg, str) or linear_pkg not in ['sklearn','scipy']:
+        raise ValueError('Attribute linear_pkg must be one of "scipy" or "sklearn".')
+
+    # select sklearn estimator
+    dict_estimators = {'Linear': LinearRegression(), 'Theil-Sen':TheilSenRegressor(random_state=random_state)
+        , 'RANSAC': RANSACRegressor(random_state=random_state), 'Huber': HuberRegressor()}
+    est = dict_estimators[estimator]
+
+    # TODO: this should be a function outside (waiting for Amaury's PR)
+    # remove NaNs
+    mykeep = np.logical_and.reduce((np.isfinite(y), np.isfinite(x)))
+    x = x[mykeep]
+    y = y[mykeep]
+
+    # subsample
+    index_sub = subsample_raster(x, 25000, return_indices=True)
+    x = x[index_sub]
+    y = y[index_sub]
+
+    # initialize cost function and output coefficients
+    mycost = np.empty(max_order)
+    coeffs = np.zeros((max_order, max_order + 1))
+    # loop on polynomial degrees
+    for deg in np.arange(1, max_order + 1):
+        # if method is linear, and package is scipy
+        if estimator == 'Linear' and linear_pkg == 'scipy':
+            # define the residual function to optimize
+            def fitfun_polynomial(xx, params):
+                return sum([p * (xx ** i) for i, p in enumerate(params)])
+            def errfun(p, xx, yy):
+                return fitfun_polynomial(xx, p) - yy
+            p0 = np.polyfit(x, y, deg)
+            myresults = scipy.optimize.least_squares(errfun, p0, args=(x, y), **kwargs)
+            if verbose:
+                print("Initial Parameters: ", p0)
+                print("Polynomial degree - ", deg, " --> Status: ", myresults.success, " - ", myresults.status)
+                print(myresults.message)
+                print("Lowest cost:", myresults.cost)
+                print("Parameters:", myresults.x)
+            mycost[deg - 1] = myresults.cost
+            coeffs[deg - 1, 0:myresults.x.size] = myresults.x
+        # otherwise, it's from sklearn
+        else:
+            if not _has_sklearn:
+                raise ValueError("Optional dependency needed. Install 'scikit-learn'")
+            p = PolynomialFeatures(degree=deg)
+            model = make_pipeline(p, est)
+            model.fit(x.reshape(-1,1), y)
+            mad = cost_func(model.predict(x.reshape(-1,1)), y)
```

Comment:
This is not necessarily MAD, right? Perhaps it could be called `error`?

Reply:
Indeed, changed into `cost` for consistency with the rest